presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java

```diff
@@ -361,6 +455,37 @@ private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expression
                 projections.build()));
     }
 
+    private List<UnfoldingExpression> prepareUnfoldingExpressions(List<SelectExpression> selectExpressions, List<Expression> symbolReferences)
+    {
+        ImmutableList.Builder<UnfoldingExpression> unfoldingExpressions = ImmutableList.builder();
+        for (int i = 0; i < selectExpressions.size(); i++) {
+            SymbolReference symbolReference = (SymbolReference) symbolReferences.get(i);
+            SelectExpression selectExpression = selectExpressions.get(i);
+            if (selectExpression.getTypeToUnfold().isPresent()) {
+                // expression to unfold of type Row: add field-reference expressions for all fields of the row
+                RowType type = selectExpression.getTypeToUnfold().get();
+                List<Type> unfoldedTypes = type.getTypeParameters();
+                for (int field = 0; field < unfoldedTypes.size(); field++) {
+                    unfoldingExpressions.add(new UnfoldingExpression(
+                            new SubscriptExpression(symbolReference, new LongLiteral("" + (field + 1))),
```

Comment:
Why do we need to do this? We're already building the dereference expressions during analysis, so we should just have to keep track of them and pass them to the planner. Maybe SelectExpressions could be refactored to do that. 

I think we can also simplify how we add the projections are being added. Let me play around with the code a bit.

Reply:
Right! If we project the whole Row expression, and then add projections for field reference expressions on top of that, they should pick up the preprojected symbol for subscript base. That simplifies things a lot. Let me refactor that.